server/rooms.ts

```diff
@@ -1253,6 +1253,72 @@ export class GlobalRoomState {
 		} catch (e) {}
 		this.lastBattle = Number(lastBattle) || 0;
 		this.lastWrittenBattle = this.lastBattle;
+		void this.loadBattles();
+	}
+
+	async saveBattles() {
+		let count = 0;
+		if (!Config.usepostgres) return 0;
+		const logDatabase = new PostgresDatabase();
+		for (const room of Rooms.rooms.values()) {
+			if (!room.battle || room.battle.ended) continue;
+			room.battle.frozen = true;
+			const log = await room.battle.getLog();
+			const players: ID[] = room.battle.options.players || [];
+			if (!players.length) {
+				for (const num of ['p1', 'p2', 'p3', 'p4'] as const) {
+					if (room.battle[num]?.id) {
+						players.push(room.battle[num].id);
+					}
+				}
+			}
+			if (!players.length || !log?.length) continue; // ???
+			await logDatabase.query(
+				`INSERT INTO stored_battles (roomid, input_log, players, title) VALUES ($1, $2, $3, $4)` +
+				` ON CONFLICT (roomid) DO UPDATE SET input_log = $2, players = $3, title = $4`,
+				[room.roomid, log, players, room.title]
+			);
+			count++;
+		}
+		return count;
+	}
+
+	async loadBattles() {
+		if (!Config.usepostgres) return;
+		const logDatabase = new PostgresDatabase();
+		const query = `DELETE FROM stored_battles WHERE roomid IN (SELECT roomid FROM stored_battles LIMIT 1) RETURNING *`;
+		for await (const battle of logDatabase.stream(query)) {
+			const {input_log, players, roomid, title, seed} = battle;
+			const [, formatid] = roomid.split('-');
+			const room = Rooms.createBattle({
+				seed,
+				format: formatid,
+				inputLog: input_log.join('\n'),
+				roomid,
+				title,
+				players,
+				delayedStart: true,
+			});
+			if (!room) continue; // ???
+			room.battle?.start();
+			for (const [i, p] of players.entries()) {
+				room.auth.set(p, Users.PLAYER_SYMBOL);
+				const u = Users.getExact(p);
+				if (u) {
+					room.battle?.joinGame(u, `p${i + 1}` as SideID);
```

Comment:
How about instead of doing `room.battle?` more than once, you just `continue` if `!room || !room.battle`